types/layer.go

```diff
@@ -379,7 +413,30 @@ func getImportFromInterface(v interface{}) (Import, error) {
 		} else {
 			hash = fmt.Sprintf("%v", m2["hash"])
 		}
-		return Import{Hash: hash, Path: fmt.Sprintf("%v", m2["Path"])}, nil
+
+		if m2["dest"] != nil {
+			if !filepath.IsAbs(m2["dest"].(string)) {
+				return Import{}, errors.Errorf("Dest path cannot be relative for: %#v", v)
+			}
+
+			dest = fmt.Sprintf("%v", m["dest"])
+		} else {
+			dest = ""
+		}
+
+		if m2["mode"] != nil {
+			mode = fs.FileMode(m2["mode"].(int))
+		}
+
+		if m2["uid"] != nil {
+			uid = m2["uid"].(int)
+		}
+
+		if m2["gid"] != nil {
+			gid = m2["gid"].(int)
+		}
+
+		return Import{Hash: hash, Path: fmt.Sprintf("%v", m2["path"]), Dest: dest, Mode: &mode, Uid: &uid, Gid: &gid}, nil
```

Comment:
I'm getting a little lost in the flow here, but this doesn't seem right.  Here, and above, you return Import{} with Mode: &mode, Uid: &uid, Gid: &gid.  But at the top of this scope, mode uid and gid are simply their zero values.  I don't see where you interpret zero values as meaning "not provided".